tests/test_server.py

```diff
@@ -85,16 +85,97 @@ def test_list_inventory(self):
     resp = self.app.get('/inventory/products')
     self.assertEqual(resp.status_code, server.HTTP_200_OK)
 
-  def test_get_pet(self):
+  def test_get_products(self):
     resp = self.app.get('/inventory/products')
     self.assertEqual(resp.status_code, server.HTTP_200_OK)
     data = json.loads(resp.data)
     self.assertEqual(int(data[0][RESTOCK_LEVEL].encode("utf-8")), 11)
 
-  def test_get_pet_not_found(self):
+  def test_get_product_not_found(self):
     resp = self.app.get('/inventory/products/0')
     self.assertEqual(resp.status_code, server.HTTP_404_NOT_FOUND)
 
+  def helper_add_new_product(self):
+    new_product = {RESTOCK_LEVEL: 20}
+    data = json.dumps(new_product)
+    return self.app.post('/inventory/products', data=data, content_type='application/json')
+
+  def helper_update_product(self):
+    updated_product = {TYPE: OPEN_BOX, QUANTITY: 12}
+    updated_data = json.dumps(updated_product)
+    id = len(server.inventory.get_all())
+    return self.app.put('/inventory/products/' + str(id), data=updated_data, content_type='application/json')
+
+  def test_update_product(self):
+    resp=self.helper_add_new_product()
+    self.assertEqual(resp.status_code, server.HTTP_201_CREATED)
+
+    resp=self.helper_update_product()
+    self.assertEqual(resp.status_code, server.HTTP_200_OK)
+
+  def test_update_product_with_exceeding_quantity(self):
+    resp = self.helper_add_new_product()
+    self.assertEqual(resp.status_code, server.HTTP_201_CREATED)
+
+    resp = self.helper_update_product()
+    self.assertEqual(resp.status_code, server.HTTP_200_OK)
+
+    id = len(server.inventory.get_all())
+
+    updated_product = {TYPE: OPEN_BOX, QUANTITY: 40}
+    updated_data = json.dumps(updated_product)
+    resp = self.app.put('/inventory/products/' + str(id), data=updated_data, content_type='application/json')
+    self.assertEqual(resp.status_code, server.HTTP_400_BAD_REQUEST)
+
+  def test_update_product_with_negative_quantity(self):
+    resp = self.helper_add_new_product()
+    self.assertEqual(resp.status_code, server.HTTP_201_CREATED)
+
+    resp = self.helper_update_product()
+    self.assertEqual(resp.status_code, server.HTTP_200_OK)
+
+    id = len(server.inventory.get_all())
+
+    updated_product = {TYPE: OPEN_BOX, QUANTITY: -1}
+    updated_data = json.dumps(updated_product)
+    resp = self.app.put('/inventory/products/' + str(id), data=updated_data, content_type='application/json')
+    self.assertEqual(resp.status_code, server.HTTP_400_BAD_REQUEST)
+
+  def test_update_product_with_invalid_data_with_missing_field(self):
+    resp = self.helper_add_new_product()
+    self.assertEqual(resp.status_code, server.HTTP_201_CREATED)
```

Comment:
Okok, but I will keep the helper because I figured it's better to add a new product(so we don't make any assumption) each time and then test it..